src/registry/artifactory.rs

```diff
@@ -1,36 +1,50 @@
 // (c) Copyright 2023 Helsing GmbH. All rights reserved.
 
-use std::sync::Arc;
-
 use eyre::{ensure, Context, ContextCompat};
-use serde::{Deserialize, Serialize};
 use url::Url;
 
 use super::Registry;
-use crate::{manifest::Dependency, package::Package};
+use crate::{
+    credentials::Credentials,
+    manifest::{Dependency, RegistryUri, Repository},
+    package::Package,
+};
 
 /// The registry implementation for artifactory
 #[derive(Debug, Clone)]
-pub struct Artifactory(Arc<ArtifactoryConfig>);
+pub struct Artifactory {
+    credentials: Credentials,
+    registry: RegistryUri,
+}
 
 impl Artifactory {
+    pub fn new(credentials: Credentials, registry: RegistryUri) -> Self {
+        Self {
+            credentials,
+            registry,
+        }
+    }
+
     /// Pings artifactory to ensure registry access is working
     pub async fn ping(&self) -> eyre::Result<()> {
-        let repositories_uri: Url = {
-            let mut url = self.0.url.clone();
-            url.set_path(&format!("{}/api/repositories", url.path()));
-            url
+        let token = self
+            .credentials
+            .get(&self.registry)
+            .wrap_err("no token found for the given registry")?;
+
+        let repositories_uri = {
+            let mut uri = self.registry.to_owned();
+            let path = &format!("{}/api/repositories", uri.path());
+            uri.set_path(path);
+            uri
         };
 
         let response = reqwest::Client::builder()
             .redirect(reqwest::redirect::Policy::none())
             .build()
             .wrap_err("client error")?
-            .get(repositories_uri.clone())
-            .header(
-                "X-JFrog-Art-Api",
-                self.0.password.clone().unwrap_or_default(),
-            )
+            .get(repositories_uri.as_str())
+            .header("X-JFrog-Art-Api", &**token)
```

Comment:
this call is optional

nit: can we make the header name a constant?

Reply:
I made the header a constant, which it definitely should be. I also made it so that the header is only set when the token exists.